source/request-response/protocol_adapter.c

```diff
@@ -0,0 +1,428 @@
+        aws_mem_calloc(allocator, 1, sizeof(struct aws_mqtt_protocol_adapter_5_subscription_op_data));
+
+    subscribe_data->allocator = allocator;
+    subscribe_data->callback_ref = aws_weak_ref_acquire(callback_ref);
+    aws_byte_buf_init_copy_from_cursor(&subscribe_data->topic_filter, allocator, topic_filter);
+
+    return subscribe_data;
+}
+
+static void s_aws_mqtt_protocol_adapter_5_subscription_op_data_destroy(
+    struct aws_mqtt_protocol_adapter_5_subscription_op_data *subscribe_data) {
+    aws_weak_ref_release(subscribe_data->callback_ref);
+    aws_byte_buf_clean_up(&subscribe_data->topic_filter);
+
+    aws_mem_release(subscribe_data->allocator, subscribe_data);
+}
+
+/* Subscribe */
+
+static void s_protocol_adapter_5_subscribe_completion(
+    const struct aws_mqtt5_packet_suback_view *suback,
+    int error_code,
+    void *complete_ctx) {
+    struct aws_mqtt_protocol_adapter_5_subscription_op_data *subscribe_data = complete_ctx;
+    struct aws_mqtt_protocol_adapter_5_impl *adapter = aws_weak_ref_get_reference(subscribe_data->callback_ref);
+
+    if (adapter == NULL) {
+        goto done;
+    }
+
+    bool success = error_code == AWS_ERROR_SUCCESS && suback != NULL && suback->reason_code_count == 1 &&
+                   suback->reason_codes[0] <= AWS_MQTT5_SARC_GRANTED_QOS_2;
+
+    struct aws_protocol_adapter_subscription_event subscribe_event = {
+        .topic_filter = aws_byte_cursor_from_buf(&subscribe_data->topic_filter),
+        .event_type = success ? AWS_PASET_SUBSCRIBE_SUCCESS : AWS_PASET_SUBSCRIBE_FAILURE,
+    };
+
+    (*adapter->config.subscription_event_callback)(&subscribe_event, adapter->config.user_data);
+
+done:
+
+    s_aws_mqtt_protocol_adapter_5_subscription_op_data_destroy(subscribe_data);
+}
+
+int s_aws_mqtt_protocol_adapter_5_subscribe(void *impl, struct aws_protocol_adapter_subscribe_options *options) {
+    struct aws_mqtt_protocol_adapter_5_impl *adapter = impl;
+
+    struct aws_mqtt_protocol_adapter_5_subscription_op_data *subscribe_data =
+        s_aws_mqtt_protocol_adapter_5_subscription_op_data_new(
+            adapter->allocator, options->topic_filter, adapter->callback_ref);
+
+    struct aws_mqtt5_subscription_view subscription_view = {
+        .qos = AWS_MQTT5_QOS_AT_LEAST_ONCE,
+        .topic_filter = options->topic_filter,
+    };
+
+    struct aws_mqtt5_packet_subscribe_view subscribe_view = {
+        .subscriptions = &subscription_view,
+        .subscription_count = 1,
+    };
+
+    struct aws_mqtt5_subscribe_completion_options completion_options = {
+        .ack_timeout_seconds_override = options->ack_timeout_seconds,
+        .completion_callback = s_protocol_adapter_5_subscribe_completion,
+        .completion_user_data = subscribe_data,
+    };
+
+    if (aws_mqtt5_client_subscribe(adapter->client, &subscribe_view, &completion_options)) {
+        goto error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
```

Comment:
On subscribe, unsubscribe, and publish, the goto error results in the completion callback not being triggered. It's probably fine to let the Request Response client timeout the operations but we could also trigger the callback here and let them know right away that something went wrong. I forget off-hand how long the timeouts were but might be nice?

Reply:
synchronous failures never invoke callbacks